modules/tinymce/src/plugins/lists/main/ts/actions/ToggleList.ts

```diff
@@ -59,6 +80,42 @@ const getEndPointNode = (editor: Editor, rng: Range, start: Boolean, root: Node)
     container = container.nextSibling;
   }
 
+  if (start && NodeType.isTextNode(container)) {
```

Comment:
So this is a lot of code targeting what looks like a relatively specific case. Often, this is what code often starts as, but ideally, it's better if we can make it more general. So here are some questions I'd ask myself as I was going through the process of getting this code ready for merging:

i) "why" are we doing this? What exactly is it trying to achieve. Could an explanatory comment about "why" help here? 
ii) is this handling the general case? Is this making assumptions about the DOM structure? If so, what are those assumptions based on? 
iii) if we are traversing left with previous sibling, I'd normally expect us to descend with `lastChild`, not `firstChild`. Is there a reason that we have picked `firstChild` for both previous and next traversals? 
iv) a lot of existing code in TinyMCE uses reassignment and lots of `if` statements, but ultimately, we want to get away from doing that. Could we rewrite this in a more intuitive way? Could we make it clearer how each movement of what 'container' is progresses? 
v) I think ultimately with a DOM traversal this complex, we'd want to test it in isolation. Could this be pulled out and tested separately?
vi) when using `textContent` we are essentially losing information about the DOM descendants and structure of the node. Is this a known limitation? Would this code still work if the original container had multiple empty elements, as well as a text node with a unicode zero width cursor.
vii) mixing `ifs` with `while` statements can make it even harder to follow what's going on. Can we pull these things out?
viii) when mixing Sugar and non-sugar APIs, things can be more confusing. Should they be mixed? And if so, is there a better boundary to mix them?

Reply:
I think you can use the TextSeeker api here so find a suitable text node. It will traverse the dom and find text nodes within the current block scope from a point that is what you need. It also takes case of complex things like noneditables etc.
However there are more complex situations here say there is a image instead of a text node `<td><a href="#"><img src="http://www.google.com/logos/google.jpg"></a></td>` there are internal apis that would find a better point here however they are not publicly exposed so the list plugin can't use them but maybe they should be though the editor.selection api so that you can do editor.selection.findNextCaretPosition etc those apis does only check for the next character in the logical order not the visual order so it might not always do what you expect in bidi text but in this case where it goes from a zwnbsp spot to the beginning of a text node they should work.

Other possible solutions could be to have the inline boundaries remove the zwnbsp point spot before the insert list  command is executed some how either by listening to the plugin command or having some public core api to do it. Having it listening for list commands makes the core dependent on the plugin in a way we usually try avoid however sometimes that is a better and easier solution also in this case the plugin is a mono repo plugin and possibly even a candidate for being moved into the core.